substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debuginfo/dwarf/DwarfSectionImpl.java

```diff
@@ -282,8 +280,8 @@ protected int putRelocatableCodeOffset(long l, byte[] buffer, int p) {
         /*
          * Mark address so it is relocated relative to the start of the text segment.
          */
-        markRelocationSite(pos, ObjectFile.RelocationKind.DIRECT_8, DwarfSectionName.TEXT_SECTION.value(), l);
-        pos = writeLong(0, buffer, pos);
+        markRelocationSite(pos, ObjectFile.RelocationKind.DIRECT_8, dwarfSections.textSectionName().value(), l);
+        pos = writeLong(dwarfSections.relocatableLong(l), buffer, pos);
```

Comment:
I'm not entirely clear why you are writing a long value at the current byte position rather than 0. Normally this content is written as zero and gets initialized with `start_address(".text") + l` when the DWARF section is loaded into memory. That happens because the associated debug reloc section identifies this location as requiring an 8-byte  symbol+offset relocation.

It looks like for MACH-O you are trying to populate this location with an absolute code address that select the correct instruction in the text section. Is that correct? Also, are you doing that that because you don't have any way of associating a relocation with this location? or is it because llvm will not apply a relocation correctly?

I am asking because it appears that for this to work you are relying on modifying all code offsets that come into the model by adding 2^32 + PAGE_SIZE. That would make sense if the text section starts at 2^32 + PAGE_SIZE (which certainly seems to fit with the name PAGEZERO_SIZE you have chosen for the basic offset).

The problem is that this trick may work to ensure that relocatable code offset locations end up with a correct instruction address but it is not going to work to generate correct heap addresses or dwarf section addresses which can also appear in DWARF content and need to be relocated. If you look at the two methods which follow, `putRelocatableHeapOffset` and `putRelocatableDwarfSectionOffset` they mark a location in the DWARF data as requiring relocation relative to either the symbol defined by `HEAP_BEGIN_NAME` or the symbol which identifies the start of a specific DWARF section. How are you proposing to ensure those offsets are generated correctly?

Reply:
Also, if you really want to transform the 4 byte offsets to 8 byte absolute addresses then you should probably be doing it here.